apps/emqx_utils/src/emqx_template.erl

```diff
@@ -346,7 +346,7 @@ render_deep({tuple, Template}, Context, Opts) when is_list(Template) ->
     {list_to_tuple(Term), Errors};
 render_deep(Template, Context, Opts) when is_list(Template) ->
     {String, Errors} = render(Template, Context, Opts),
-    {unicode:characters_to_binary(String), Errors};
+    {character_segments_to_binary(String), Errors};
```

Comment:
This is the trickiest part.
Currently, `render_str` _allows_ to interpolate binary non-utf8 data into templates, and this is widely used, e.g., in interpolating SQL parameters.

```erlang
> emqx_template:render(emqx_template:parse(<<"hello ${username}">>), #{username => <<255, 255, 255>>}).
{[<<"hello ">>,<<"ÿÿÿ">>],[]}
```

But `render_deep` does not allow interpolating binary non-utf8 data, and in a hidden way, some nested values become `{error, ..., ...}` tuple, thus silently making rendered structure invalid (e.g. for encoding into json).

```erlang
> emqx_template:render(emqx_template:parse_deep(#{username => <<"${username}">>}), #{username => <<255, 255, 255>>}).
{#{username => {error,<<>>,<<"ÿÿÿ">>}},[]}
```

Since the produced error cannot be used anyhow and is obviously silently ignored (m.b. resulting in some crashes), we make the rendering more graceful: we do not reencode binaries and reencode non-encodable lists as json.

Unicode enforcement may be shifted to the `var_trans` part, see e.g. `emqx_auth_utils:render_deep_for_raw/2` vs `emqx_auth_utils:render_deep_for_json/2`